Inst synthesis

rsas opened the pull request:
The synthesis code base grows and I think it's time to review this patch so that others also understand the core algorithm and its peculiarities. Please comment on everything that is unclear so I can improve both the code and the documentation. Please read the commit message from 8b5e5fd that contains instructions for synthesis invocation and usage. The main difference from that commit is that now there is no API and the synthesis is iterative, i.e., it starts with zero components (nop synthesis) and proceeds until it reaches the max num of components specified via cmd or the LHS cost minus one. If no max components were specified, it will try to use the whole component library (or all components specified via cmd).

rsas commented:
I also highly encourage to watch the synthesis in action using the ``-souper-debug-synthesis`` flag. For examples, have a look at test cases in ``test/Infer/*-syn.*`` and ``test/Infer/HackersDelight``.

buildhive commented:
[Google » souper #462](https://buildhive.cloudbees.com/job/google/job/souper/462/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

chenyang78 commented:
Fascinating! One silly question, what's nop synthesis? Thanks.

regehr commented:
Yang, nop synthesis is what we call it when Souper finds an optimization that doesn't require adding any instructions at all because the required value is already available. For example:

```
%0:i32 = var
%1:i32 = lshr %0, 16:i32
%2:i64 = zext %1
%3:i64 = and 255:i64, %2
%4:i32 = lshr %0, 24:i32
%5:i64 = zext %4
%6:i64 = shlnw %5, 8:i64
%7:i64 = or %3, %6
infer %7
result %2
```

rsas commented:
@regehr nitpicking: The synthesis in the example above found a sequence of two instructions that was in the original LHS too. For me, nop synthesis is rather the following opt where one of the inputs is wired to the output directly:

```
%0:i32 = var
%1:i64 = var
%2:i64 = lshr %1, 3:i64
%3:i32 = trunc %2
%4:i1 = ult %0, %3
%5:i32 = sub %0, %3
%6:i32 = select %4, 0:i32, %5
%7:i1 = slt 0:i32, %6
pc %7 1:i1
%8:i1 = ult %3, %0
%9:i32 = select %8, %0, %3
infer %9
result %0
```
In the synthesis code, this is how nop synthesis is defined, i.e., find an opt using zero components.

regehr commented:
Well... I'm not sure it matters whether the result refers to a var or to an instruction -- either way, some LLVM instructions have been elided, as opposed to being expressed using a different instruction sequence.

regehr commented on include/souper/Infer/InstSynthesis.h:
```diff
@@ -0,0 +1,300 @@
+/// 4_0 describes the final output (assuming there are only three components)
+typedef std::pair<unsigned, unsigned> LocVar;
+
+/// A mapping from a location variable to it's concrete value (Inst::Const).
```
http://www.its-not-its.info/

rsas replied:
Argh, @jeroenk pointed to a number of these already. I thought I got rid of these, but obviously not.

rsas commented:
@regehr ok, let's agree then when a sequence of synthesized instructions is in the LHS, it's a nop. It can be found using 0 or more components.

regehr commented on lib/Extractor/Solver.cpp:
```diff
@@ -111,6 +115,13 @@ class BaseSolver : public Solver {
+    if (InferInsts && SMTSolver->supportsModels()) {
+      InstSynthesis IS;
+      EC = IS.synthesize(SMTSolver.get(), BPCs, PCs, LHS, RHS, IC, Timeout);
+      if (EC || RHS)
```
Please make sure it is OK to read RHS here (that is, make sure synthesize() has set it to 0 when synthesis failed. I ran into a bug of this kind recently.

rsas replied:
Thanks, I ran into a bug recently too and now I'm setting RHS to 0 explicitly since then.

regehr commented on lib/Extractor/Solver.cpp:
```diff
@@ -111,6 +115,13 @@ class BaseSolver : public Solver {
+        return EC;
+    }
+
     RHS = 0;
```
If synthesize() has set RHS to 0 when synthesis fails, this is unnecessary code.

rsas replied:
Yes, but we still have to cover the path where synthesis is not invoked at all, right?

regehr commented on lib/Infer/InstSynthesis.cpp:
```diff
@@ -0,0 +1,1264 @@
+
+  // Ask the solver for a couple of arbitrary initial inputs
+  auto InputPCs = PCs;
+  for (unsigned I = 0; I < 4; ++I) {
```
Where does 4 come from?
Be careful about variables named I since we often use that to mean an Inst *.

regehr commented on lib/Infer/InstSynthesis.cpp:
```diff
@@ -0,0 +1,1264 @@
+  for (unsigned I = 0; I < 4; ++I) {
+    std::vector<Inst *> ModelInsts;
+    std::vector<llvm::APInt> ModelVals;
+    InstMapping Mapping(LHS, IC.createVar(LHS->Width, "foo"));
```
Is there a better name than foo?

regehr commented on lib/Infer/InstSynthesis.cpp:
```diff
@@ -0,0 +1,1264 @@
+  if (MaxCompNum > (int)Comps.size())
+    MaxCompNum = Comps.size();
+  // Iterative synthesis loop with increasing number of components
+  for (int J = 0; J <= MaxCompNum; ++J) {
```
earlier J was unsigned, is there a reason it is signed here?

rsas replied:
Yes, ``MaxCompNum`` becomes negative for the following LHS with cost 0:

```
%0:i8 = var
infer %0
```

MaxCompNum = LHSCost-1;
=> MaxCompNum = -1;

In this case, we don't want to enter the synthesis loop at all.

regehr commented on lib/Infer/InstSynthesis.cpp:
```diff
@@ -0,0 +1,1264 @@
+      Inst *Cand = createInstFromModel(std::make_pair(ModelInsts, ModelVals),
+                                       Wiring, IC);
+      if (!Cand) {
+        llvm::errs() << "synthesis bug: creating inst from a model failed\n";
```
If this truly should never happen, I'd prefer to crash souper when it does, instead of printing something.

regehr commented on lib/Infer/InstSynthesis.cpp:
```diff
@@ -0,0 +1,1264 @@
+    if (In->Width > DefaultWidth)
+      DefaultWidth = In->Width;
+  // No inputs, use output width
+  if (!DefaultWidth)
```
Can you explain how no inputs can happen? Wouldn't any Souper expression have at least one input?

rsas replied:
It happened when I was running synthesis on a test case derived from test/Infer/bswap.ll:

```
define i32 @foo(i32 %x) {
entry:
  %swap = call i32 @llvm.bswap.i32(i32 2882343476) ;input is 0xABCD1234
  ret i32 %swap ;swapped result is 0x3412CDAB
}
```
LHS %swap has no inputs.

regehr commented on include/souper/Infer/InstSynthesis.h:
```diff
@@ -0,0 +1,300 @@
+/// Supported component library.
+/// The width of most of the instructions is unknown in advance, therefore, we
+/// indicate this by setting the Width to 0. During initialization, the width
+/// will be set to the DefaultWidth (maximum width of the input vars).
```
Do we care about mixed-width expressions? When looking at the stuff that comes from gzip, I seem to see a lot of expressions that mix different widths. Is it too expensive to instantiate components at every width that might be useful?

rsas replied:
At the moment, definitely too expensive. Let's consider this once the performance of synthesis significantly improves.

regehr commented on lib/Infer/InstSynthesis.cpp:
```diff
@@ -0,0 +1,1264 @@
+  }
+
+  if (!CompInstMap.count(OutLoc)) {
+    llvm::errs() << "synthesis bug: output not wired\n";
```
report_fatal_error()?

regehr commented on lib/Infer/InstSynthesis.cpp:
```diff
@@ -0,0 +1,1264 @@
+    LocVar Match = getWiringLocVar(OpLoc, ProgramWiring);
+    assert(CompInstMap.count(Match) && "unknown matching location variable");
+    if (!CompInstMap.count(Match)) {
+      llvm::errs() << "synthesis bug: component input "
```
report_fatal_error()?

regehr commented on lib/Infer/InstSynthesis.cpp:
```diff
@@ -0,0 +1,1264 @@
+                                        InstContext &IC) {
+  switch (Kind) {
+  case Inst::Add:
+  case Inst::AddNSW:
```
But aren't the nsw/nuw variants in the list of components not supported by synthesis above?

rsas replied:
They are not supported yet (scalability), but once they are, this function will handle them properly.

regehr commented on lib/Infer/InstSynthesis.cpp:
```diff
@@ -0,0 +1,1264 @@
+  }
+
+  case Inst::And:
+  case Inst::Or: {
```
Do we care to look for and/or with all 1s or all 0s?

rsas replied:
Isn't that already captured by the lines below?

regehr commented:
Well generally this looks great. Since it has minimal impact on Souper code that runs in non-synthesis mode, I don't see any reason not to take this patch. It would be good for @pcc and @chenyang78 to look it over as well. 

Once this lands, I'll revive my synthesis code in the pass. I do not believe that will be a huge amount of work.

rsas commented:
Thanks for the review @regehr, I'll address them on Monday.

pcc commented on CMakeLists.txt:
```diff
@@ -242,13 +252,14 @@ target_link_libraries(souper souperExtractor souperKVStore souperParser souperSM
 target_link_libraries(parser_tests souperParser ${GTEST_LIBS})
 
 set(TEST_SOLVER "" CACHE STRING "Solver command line argument to use in tests, e.g. -stp-path=/path/to/stp")
+set(TEST_SYNTHESIS "" CACHE STRING "Enable program synthesis tests")
```
Why are these tests disabled by default? I would strongly prefer that have at least some test coverage for all features by default without a good reason otherwise.

rsas replied:
The only reason is testing runtime that is typically longer during synthesis. If we don't care about that, I can enable these test cases by default.

regehr replied:
Is a compromise available where for example you run synthesis with a very limited set of components?

rsas replied:
Yes, I can remove a couple of Hacker's Delight examples. Nevertheless, I still want to have a big synthesis test suite for both testing and performance measurements.

rsas replied:
With the current set of 29 synthesis test cases, the testing time using four cores on my machine goes up from 15.42s to 28.20s. If you are OK with that, I'll enable the test cases by default.

regehr replied:
It sounds like testing time will be acceptable if you move a subset into the default test config -- try to pick a collection that together stress the important features.

rsas replied:
I'm moving 13 test cases to the default test suite while leaving others, i.e., computationally intensive ones, optional.

buildhive commented:
[Google » souper #463](https://buildhive.cloudbees.com/job/google/job/souper/463/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

buildhive commented:
[Google » souper #464](https://buildhive.cloudbees.com/job/google/job/souper/464/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

buildhive commented:
[Google » souper #465](https://buildhive.cloudbees.com/job/google/job/souper/465/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

buildhive commented:
[Google » souper #466](https://buildhive.cloudbees.com/job/google/job/souper/466/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

buildhive commented:
[Google » souper #467](https://buildhive.cloudbees.com/job/google/job/souper/467/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

buildhive commented:
[Google » souper #468](https://buildhive.cloudbees.com/job/google/job/souper/468/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

rsas commented:
Thanks for the comments @chenyang78 @pcc @regehr, I addressed all of them. If you have time, please have a look. If there are no further objections, we should merge it ASAP so that @regehr can add the Pass support.

chenyang78 commented:
I don't have any concern at the moment. Again, thanks for putting this together, Raimondas and John.

rsas commented:
Thanks @chenyang78. @regehr hit a tricky bug yesterday, that I haven't seen so far. I'll have to do more testing.

buildhive commented:
[Google » souper #469](https://buildhive.cloudbees.com/job/google/job/souper/469/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

rsas commented:
@regehr I addressed the bug, you can start testing. Please let me know once you are LGTM with the patch, so I can merge.

buildhive commented:
[Google » souper #470](https://buildhive.cloudbees.com/job/google/job/souper/470/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

buildhive commented:
[Google » souper #471](https://buildhive.cloudbees.com/job/google/job/souper/471/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

rsas commented:
@regehr I squashed the commits and ran tests in Debug/Release modes. Please have a final look and merge.

regehr commented on include/klee/Config/config.h:
```diff
@@ -13,4 +13,4 @@
 // limitations under the License.
 
 #define LLVM_VERSION_MAJOR 3
-#define LLVM_VERSION_MINOR 6
```
The souper trunk is already at LLVM_VERSION_MINOR 7 since the last bump. Raimonads could you make sure that your patch here is against the latest master?

rsas replied:
It is. My squash includes all master commits since my initial commit in my branch. I did a test merge on my local master and it applies cleanly.

rsas replied:
I merged to master from time to time, that's why you see these lines.

rsas replied:
Argh, I see your point now. I thought git would be smart enough to remove these existing commits automatically during merge, but it doesn't. I'll have to fix it first.

regehr commented:
I am getting this from a Debug build:

```
regehr@ulam:build{inst-synthesis}$ make check
[  2%] Built target profileRuntime
[ 45%] Built target souperPass
[ 47%] Built target souperSMTLIB2
[ 50%] Built target souperInfer
[ 52%] Built target souperInst
[ 75%] Built target kleeExpr
[ 81%] Built target souperExtractor
[ 83%] Built target souperTool
[ 85%] Built target souperKVStore
[ 87%] Built target souperParser
[ 89%] Built target souper
[ 91%] Built target parser-test
[ 93%] Built target souper-check
[ 95%] Built target extractor_tests
[ 97%] Built target inst_tests
[100%] Built target parser_tests
FAIL: Souper :: Infer/signs-syn.opt (70 of 151)
******************** TEST 'Souper :: Infer/signs-syn.opt' FAILED ********************
Script:
--
/home/regehr/souper-rsas/build/souper-check -z3-path=/usr/bin/z3 -infer-rhs -souper-infer-inst -souper-synthesis-comps=const,slt,xor /home/regehr/souper-rsas/test/Infer/signs-syn.opt > /home/regehr/souper-rsas/build/test/Infer/Output/signs-syn.opt.tmp1
/home/regehr/souper-rsas/third_party/llvm/Debug/bin/FileCheck /home/regehr/souper-rsas/test/Infer/signs-syn.opt < /home/regehr/souper-rsas/build/test/Infer/Output/signs-syn.opt.tmp1
--
Exit Code: 1

Command Output (stderr):
--
/home/regehr/souper-rsas/test/Infer/signs-syn.opt:6:10: error: expected string not found in input
; CHECK: %9:i32 = xor %0, %2
         ^
<stdin>:1:1: note: scanning from here
; Failed to infer RHS
^
<stdin>:1:6: note: possible intended match here
; Failed to infer RHS
     ^

--

********************
Testing Time: 70.59s
********************
Failing Tests (1):
    Souper :: Infer/signs-syn.opt

  Expected Passes    : 134
  Unsupported Tests  : 16
  Unexpected Failures: 1
make[3]: *** [CMakeFiles/check] Error 1
make[2]: *** [CMakeFiles/check.dir/all] Error 2
make[1]: *** [CMakeFiles/check.dir/rule] Error 2
make: *** [check] Error 2
regehr@ulam:build{inst-synthesis}$ 
```

rsas commented:
Weird, no failures here. z3 version?

regehr commented:
Weird, I don't know. My z3 was a bit old (unstable from last fall) but I upgraded to unstable as of right now (it calls itself 4.4.0) and still run into the same problem.

regehr commented:
Here's the raw command:

```
$ /home/regehr/souper-rsas/build/souper-check -z3-path=/usr/bin/z3 -infer-rhs -souper-infer-inst -souper-synthesis-comps=const,slt,xor /home/regehr/souper-rsas/test/Infer/signs-syn.opt
; Failed to infer RHS
```

rsas commented:
I fetched a fresh copy of my branch and ran the same command:

```
$ ./souper-check -z3-path=`which z3` -infer-rhs -souper-infer-inst -souper-synthesis-comps=const,slt,xor ../test/Infer/signs-syn.opt 
; RHS inferred successfully
%9:i32 = xor %0, %2
%10:i1 = slt %9, 0:i32
result %10
```
Z3 version 4.4.0 from unstable branch.

rsas commented:
Did you update my 'inst-synthesis' branch?

rsas commented on lib/Extractor/Candidates.cpp:
```diff
@@ -362,8 +362,10 @@ Inst *ExprBuilder::build(Value *V) {
+    // TODO: replace with a GEP instruction
+    //return buildGEP(get(GEP->getOperand(0)), gep_type_begin(GEP),
+    //                gep_type_end(GEP));
+    return makeArrayRead(V);
```
@regehr are you OK with commenting ``buildGEP`` out in the master?

regehr replied:
I think it's an OK simplifying assumption for now.

rsas commented:
I did a clean commit and a rebase to master. The only thing I'm not sure is the ``buildGEP`` part, see comment above. @regehr please do a fresh checkout of my ``inst-synthesis`` branch and run all test cases again. On my machine building and testing in Debug/Release modes didn't show any problems.

buildhive commented:
[Google » souper #472](https://buildhive.cloudbees.com/job/google/job/souper/472/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

regehr commented:
The test failure was from a fresh checkout and build of your branch on my Ubuntu 14.04 machine at work. I'm home now and will see what happens here.

regehr commented:
I see the same thing at home. You've definitely tried a Debug build?

buildhive commented:
[Google » souper #473](https://buildhive.cloudbees.com/job/google/job/souper/473/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

chenyang78 commented:
I got the same error in the Release build (haven't tried Debug version yet)

    /home/chenyang/work/souper_fork/souper_ras/build_release/souper-check -z3-path=/home/chenyang/other_programs/z3/build/z3 -infer-rhs -souper-infer-inst -souper-synthesis-comps=const,slt,xor /home/chenyang/work/souper_fork/souper_ras/test/Infer/signs-syn.opt > /home/chenyang/work/souper_fork/souper_ras/build_release/test/Infer/Output/signs-syn.opt.tmp1
    /home/chenyang/work/souper_fork/souper_ras/third_party/llvm/Release/bin/FileCheck /home/chenyang/work/souper_fork/souper_ras/test/Infer/signs-syn.opt < /home/chenyang/work/souper_fork/souper_ras/build_release/test/Infer/Output/signs-syn.opt.tmp1
    --
    Exit Code: 1
    
    Command Output (stderr):
    --
    /home/chenyang/work/souper_fork/souper_ras/test/Infer/signs-syn.opt:6:10: error: expected string not found in input
    ; CHECK: %9:i32 = xor %0, %2
             ^
    <stdin>:1:1: note: scanning from here
    ; Failed to infer RHS
    ^
    <stdin>:1:6: note: possible intended match here
    ; Failed to infer RHS
         ^

I am using z3 4.3.1.

rsas commented:
This test case is problematic, I removed it in my last squashed commit.

chenyang78 commented:
I see. Thanks!

buildhive commented:
[Google » souper #474](https://buildhive.cloudbees.com/job/google/job/souper/474/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

buildhive commented:
[Google » souper #475](https://buildhive.cloudbees.com/job/google/job/souper/475/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

regehr commented on lib/Infer/InstSynthesis.cpp:
```diff
@@ -0,0 +1,1376 @@
+
+namespace {
+
+static cl::opt<bool> DebugSynthesis("souper-synthesis-debug",
```
I think it would make sense to collapse these into a single variable where level 0 means debugging is turned off, etc.

rsas replied:
Yes, totally.

buildhive commented:
[Google » souper #476](https://buildhive.cloudbees.com/job/google/job/souper/476/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

buildhive commented:
[Google » souper #477](https://buildhive.cloudbees.com/job/google/job/souper/477/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

regehr commented on lib/Infer/InstSynthesis.cpp:
```diff
@@ -0,0 +1,1373 @@
+  case Inst::Cttz:
+  case Inst::Ctlz: {
+    if (Ops[0] == IC.getConst(APInt(Width, 0)))
+      return IC.getConst(APInt(Width, Width));
```
It's a nit, but do we want to use curlies around a case that doesn't declare any local variables?

rsas replied:
No.

regehr commented on lib/Infer/InstSynthesis.cpp:
```diff
@@ -0,0 +1,1373 @@
+
+std::string InstSynthesis::getLocVarStr(const LocVar &Loc,
+                                        const std::string Prefix) {
+  std::string Post = "";
```
a std::string is implicitly initialized to "" I believe

rsas replied:
You are right.

buildhive commented:
[Google » souper #478](https://buildhive.cloudbees.com/job/google/job/souper/478/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

buildhive commented:
[Google » souper #479](https://buildhive.cloudbees.com/job/google/job/souper/479/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

buildhive commented:
[Google » souper #480](https://buildhive.cloudbees.com/job/google/job/souper/480/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

buildhive commented:
[Google » souper #481](https://buildhive.cloudbees.com/job/google/job/souper/481/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

regehr commented on include/souper/Infer/InstSynthesis.h:
```diff
@@ -0,0 +1,300 @@
+  /// namely created instruction
+  std::map<LocVar, Inst *> CompInstMap;
+  /// Location variable's width is fixed
+  const unsigned LocInstWidth = 32;
```
Is there any reason to use so many bits here? Might it be helpful to reduce this to 6 or 8?

rsas replied:
No. I think that 6 (max 64 comps) should bu sufficient for now.

regehr commented on lib/Infer/InstSynthesis.cpp:
```diff
@@ -0,0 +1,1360 @@
+  // Ask the solver for four initial concrete inputs.
+  // This number was derived experimentally giving a good overall speed-up
+  // for both small and big synthesis queries
+  unsigned InitialInputNum = 4;
```
Is there a reason for not declaring this variable in the for-loop header below?

rsas replied:
No.

regehr commented on lib/Infer/InstSynthesis.cpp:
```diff
@@ -0,0 +1,1360 @@
+      llvm::outs() << "synthesizing using " << J << " component(s)\n";
+    // If synthesis using 0 components failed (aka nop synthesis),
+    // don't subsequently wire the output to the input(s)
+    if (J == 0)
```
please don't use curly braces on the else-branch but not on the if-branch

regehr commented on lib/Infer/InstSynthesis.cpp:
```diff
@@ -0,0 +1,1360 @@
+    if (It->Kind == Inst::BSwap && It->Width != 16 && It->Width != 32 &&
+        It->Width != 64)
+      It = Comps.erase(It);
+    else if (((It->Kind == Inst::CtPop) || (It->Kind == Inst::Ctlz) ||
```
personally I'd put more curly braces in a collection of ifs/elses that is this complicated

regehr commented on lib/Infer/InstSynthesis.cpp:
```diff
@@ -0,0 +1,1360 @@
+
+void InstSynthesis::addConstraints(std::vector<InstMapping> &PCs,
+                                   InstContext &IC) {
+  Inst *C = getConsistencyConstraint(IC);
```
repeated use of C for different purposes is a bit odd here-- would the code look better if you just inlined the calls that are generating the constraints inside the emplace_back calls?

rsas replied:
Absolutely.

regehr commented:
other than issues just noted, LGTM -- please look over them and then squash and merge

rsas commented:
@regehr thanks for the review! I addressed your comments in c7e4777. In addition, I removed ``addConstraints`` (I think it's better to have it in the main function) and introduced ``TrueConst`` in the main function to avoid redundant calls to ``IC.getConst(APInt(1, true))``. All test cases in Debug/Release modes with both KLEE versions are passing. Please have a short look at the last commit and then I'll merge it.

buildhive commented:
[Google » souper #482](https://buildhive.cloudbees.com/job/google/job/souper/482/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

regehr commented:
Looks good, please squash and merge.

buildhive commented:
[Google » souper #483](https://buildhive.cloudbees.com/job/google/job/souper/483/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)
